File: schemas/social-media/drizzle/reactions.ts
// reactions: Typed reactions (like, love, etc.) on posts.
// See README.md for full design rationale.
import { pgTable, pgEnum, uuid, timestamp, index, unique } from "drizzle-orm/pg-core";
import { posts } from "./posts";
import { users } from "./users";

export const reactionTypeEnum = pgEnum("reaction_type", ["like", "love", "celebrate", "insightful", "funny"]);

export const reactions = pgTable(
  "reactions",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    postId: uuid("post_id")
      .notNull()
      .references(() => posts.id, { onDelete: "cascade" }),
    userId: uuid("user_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),
    type: reactionTypeEnum("type").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_reactions_post_user_type").on(table.postId, table.userId, table.type),
    index("idx_reactions_user_id").on(table.userId),
  ]
);

File: schemas/social-media/spacetimedb/reactions.rs
// reactions: user reactions (like, love, etc.) on posts.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum ReactionType {
    Like, // type: String
    Love,
    Celebrate,
    Insightful,
    Funny,
}

#[spacetimedb::table(name = reactions, public)]
pub struct Reaction {
    #[primary_key]
    pub id: String, // UUID
    // Composite unique: (post_id, user_id, type) — enforce in reducer logic.
    #[index(btree)]
    pub post_id: String, // UUID — FK → posts.id (cascade delete)
    #[index(btree)]
    pub user_id: String, // UUID — FK → users.id (cascade delete)
    pub r#type: ReactionType,
    pub created_at: Timestamp,
}
